### PHASE 1:

In [1]:
import numpy as np
import pandas as pd

### 1. LOADING THE DATASET

In [2]:
dataset=pd.read_csv(r"C:\Users\mynam\Innomatics\Batch 406\water_scarcity_consumption_updated.csv")

### 2. VIEWING SAMPLE DATA

In [4]:
dataset.head()

Region    Year  Population_Million  Water_Available_MCM  \
0     West  2006.0               169.0                955.0   
1  Central  2007.0               196.0               2966.0   
2     East  2000.0                13.0                282.0   
3  Central  2022.0                49.0               3212.0   
4  Central  2002.0                60.0               5770.0   

   Water_Consumption_MCM  Agriculture_Use_MCM  Industrial_Use_MCM  \
0                 4779.0               3657.0              2501.0   
1                 1723.0               2183.0              2504.0   
2                 8356.0               4375.0               995.0   
3                    NaN               1118.0              1002.0   
4                 2344.0               5619.0              2772.0   

   Domestic_Use_MCM  Rainfall_mm  Groundwater_Level_m  Scarcity_Index  \
0              23.0        667.0                24.06            0.73   
1             382.0       2215.0                 8.71            0.05   
2             234.0       2406.0                 2.52            0.88   
3            1538.0       2197.0                47.40            0.86   
4            1106.0       2358.0                12.79            0.44   

   Per_Capita_Consumption_LPD Continent Climate_Type        Data_Source  
0                       118.0    Europe    Semi-Arid           Research  
1                       474.0    Africa         Arid  Government Report  
2                       382.0    Europe    Semi-Arid          Satellite  
3                       306.0    Europe    Semi-Arid             Survey  
4                       138.0      Asia    Semi-Arid           Research

OBSERVATION: 

This gives us the first few records of the dataset(so that we can confirm whether parsing is correct).

### 3. CHECKING THE STRUCTURE OF DATA

In [8]:
dataset.shape

(1100, 15)

OBSERVATION: 

The dataset contains 1100 rows and 15 columns.

In [5]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1100 entries, 0 to 1099
Data columns (total 15 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Region                      1064 non-null   object 
 1   Year                        1070 non-null   float64
 2   Population_Million          1063 non-null   float64
 3   Water_Available_MCM         1065 non-null   float64
 4   Water_Consumption_MCM       1067 non-null   float64
 5   Agriculture_Use_MCM         1064 non-null   float64
 6   Industrial_Use_MCM          1068 non-null   float64
 7   Domestic_Use_MCM            1062 non-null   float64
 8   Rainfall_mm                 1066 non-null   float64
 9   Groundwater_Level_m         1069 non-null   float64
 10  Scarcity_Index              1063 non-null   float64
 11  Per_Capita_Consumption_LPD  1062 non-null   float64
 12  Continent                   1100 non-null   object 
 13  Climate_Type                1100 

#### OBSERVATION: 

This command reveals:

* Column names

* Data types (int, float, object/string)

* Non-null counts (so you know which columns have missing values).

Expected:

* Many numeric Columns: Population_Million, Water_Available_MCM, Water_Consumption_MCM, Scarcity_Index, Rainfall_mm, etc.

* Several categorical/text Columns: Region, Continent, Climate_Type, Data_Source.

* Missing values across multiple columns (for example: Year, Water_Consumption_MCM, Agriculture_Use_MCM are not always present).

### 4. GETTING THE SUMMARY STATS OF DATA:

In [13]:
dataset.describe()

Year  Population_Million  Water_Available_MCM  \
count  1070.000000         1063.000000          1065.000000   
mean   2011.305607          101.529633          4977.624413   
std       6.858115           57.470690          2801.815863   
min    2000.000000            1.000000           106.000000   
25%    2005.000000           52.000000          2568.000000   
50%    2011.000000          106.000000          4881.000000   
75%    2017.000000          149.500000          7396.000000   
max    2023.000000          199.000000          9989.000000   

       Water_Consumption_MCM  Agriculture_Use_MCM  Industrial_Use_MCM  \
count            1067.000000          1064.000000         1068.000000   
mean             4378.036551          3464.177632         1491.230337   
std              2533.361794          2004.980860          889.290936   
min                52.000000            21.000000           14.000000   
25%              2164.000000          1707.000000          710.000000   
50%              4282.000000          3449.000000         1433.000000   
75%              6562.000000          5234.750000         2293.000000   
max              8974.000000          6995.000000         2996.000000   

       Domestic_Use_MCM  Rainfall_mm  Groundwater_Level_m  Scarcity_Index  \
count       1062.000000  1066.000000          1069.000000     1063.000000   
mean        1011.520716  1536.355535            24.719121        0.489464   
std          568.478037   839.982315            13.897171        0.282016   
min           15.000000    52.000000             1.010000        0.000000   
25%          537.000000   832.250000            12.760000        0.250000   
50%         1001.000000  1519.000000            24.760000        0.480000   
75%         1513.000000  2258.000000            36.000000        0.720000   
max         1999.000000  2997.000000            49.950000        1.000000   

       Per_Capita_Consumption_LPD  
count                 1062.000000  
mean                   273.697740  
std                    132.515381  
min                     50.000000  
25%                    156.000000  
50%                    278.000000  
75%                    388.000000  
max                    499.000000

OBSERVATION: 

This provides:

* The mean, std, min, max, count etc for various numerical columns.

* It helps detect anomalies (e.g., extremely large Water_Consumption_MCM values or negative ones).
We may notice inconsistent scales — e.g., Scarcity_Index is between 0–1 (a ratio), while MCM values (million cubic meters) are very large.

### 5. FINDING MISSING VALUES:

In [14]:
dataset.isnull().sum()

Region                        36
Year                          30
Population_Million            37
Water_Available_MCM           35
Water_Consumption_MCM         33
Agriculture_Use_MCM           36
Industrial_Use_MCM            32
Domestic_Use_MCM              38
Rainfall_mm                   34
Groundwater_Level_m           31
Scarcity_Index                37
Per_Capita_Consumption_LPD    38
Continent                      0
Climate_Type                   0
Data_Source                    0
dtype: int64

This will show how many NaN (missing) entries exist per column.

Likely issues:

* Year: missing in some rows

* Population_Million: missing

* Industrial_Use_MCM, Domestic_Use_MCM, etc. — many blank values

* Some categorical columns (like Region) are empty in certain rows



In [15]:
round(dataset.isnull().sum()/len(dataset)*100,2)

Region                        3.27
Year                          2.73
Population_Million            3.36
Water_Available_MCM           3.18
Water_Consumption_MCM         3.00
Agriculture_Use_MCM           3.27
Industrial_Use_MCM            2.91
Domestic_Use_MCM              3.45
Rainfall_mm                   3.09
Groundwater_Level_m           2.82
Scarcity_Index                3.36
Per_Capita_Consumption_LPD    3.45
Continent                     0.00
Climate_Type                  0.00
Data_Source                   0.00
dtype: float64

OBSERVATION:

* Most complete columns → Data_Source, Climate_Type, Continent (almost no missing values).

* Highest missing values → Domestic_Use_MCM (1.41%), followed by Industrial_Use_MCM (1.34%) and Agriculture_Use_MCM (1.19%).

* On average, ~0.8–1.5% of the dataset is missing across columns.

### 6. UNDERSTANDING CATEGORIES:

`For Categorical columns:`

a) Region

* Info: Name of the sub-region (e.g., West, Central, North, etc.)

* Use: Helps segment water data geographically and compare across regions.

* Missing Values: ~70+ rows with missing Region.

In [19]:
print(dataset["Region"].value_counts())

Region
North      230
West       215
Central    212
South      206
East       201
Name: count, dtype: int64


b) Continent

* Info: Continent to which the region belongs.

* Use: Enables regional grouping & analysis on broader scale.

* Missing Values: Very few (≤10 cases).

In [20]:
print(dataset["Continent"].value_counts())

Continent
Asia             237
South America    231
Africa           230
North America    208
Europe           194
Name: count, dtype: int64


c) Climate_Type

* Info: Climate classification (Arid, Semi-Arid, Temperate, Tropical, etc.).

* Use: Allows climate–water relationship study e.g., rainfall vs scarcity.

* Missing Values: ~20 rows.


In [21]:
print(dataset["Climate_Type"].value_counts())


Climate_Type
Tropical     290
Semi-Arid    275
Arid         270
Temperate    265
Name: count, dtype: int64


d) Data_Source

* Info: Origin of data (Survey, Government Report, Satellite, Research).

* Use: Assesses reliability, traceability, and uniformity of the dataset.

* Missing Values: ~10 rows.

In [22]:
print(dataset["Data_Source"].value_counts())

Data_Source
Research             292
Survey               286
Government Report    263
Satellite            259
Name: count, dtype: int64


`For Numerical Columns:`

a) Year:

* Info: Year of observation/recording (2000–2023 mostly).

* Use: Enables time-series analysis to track trends in population, water use, scarcity over years.

* Missing Values: ~50 missing entries.

In [26]:
print(dataset["Year"].value_counts())

Year
2014.0    54
2001.0    53
2011.0    52
2016.0    51
2017.0    50
2012.0    48
2002.0    48
2013.0    47
2006.0    46
2004.0    46
2007.0    45
2003.0    45
2009.0    45
2010.0    45
2020.0    44
2023.0    44
2021.0    42
2022.0    42
2005.0    41
2008.0    41
2000.0    39
2018.0    35
2019.0    34
2015.0    33
Name: count, dtype: int64


b) Population_Million:

* Info: Regional population in millions during that year.

* Use: Needed to calculate per capita consumption and demand-supply ratios.

* Missing Values: ~80 entries.

In [27]:
print(dataset["Population_Million"].value_counts())

Population_Million
149.0    17
86.0     11
120.0    11
21.0     10
12.0     10
         ..
83.0      2
40.0      2
18.0      1
123.0     1
53.0      1
Name: count, Length: 199, dtype: int64


c) Water_Available_MCM:

* Info: Total water available in Million Cubic Meters (MCM) annually.

* Use: Measures supply/capacity. Crucial for balance vs. consumption.

* Missing Values: ~40 rows.

In [30]:
print(dataset["Water_Available_MCM"].value_counts())

Water_Available_MCM
982.0     3
5768.0    3
2706.0    3
9356.0    2
2046.0    2
         ..
9556.0    1
7261.0    1
1984.0    1
1909.0    1
7217.0    1
Name: count, Length: 1007, dtype: int64


d) Water_Consumption_MCM:

* Info: Total annual consumption of water in MCM.

* Use: Key metric to assess demand and scarcity when compared with availability.

* Missing Values: ~70 rows.

In [31]:
print(dataset["Water_Consumption_MCM"].value_counts())

Water_Consumption_MCM
2631.0    3
665.0     3
1931.0    3
4528.0    2
5212.0    2
         ..
894.0     1
496.0     1
3965.0    1
3159.0    1
4263.0    1
Name: count, Length: 1003, dtype: int64


e) Agriculture_Use_MCM: 

* Info: Water consumption for agriculture in MCM.

* Use: Agriculture is usually the largest water consumer. Essential for sustainable planning.

* Missing Values: ~90 rows.

In [32]:
print(dataset["Agriculture_Use_MCM"].value_counts())

Agriculture_Use_MCM
1482.0    4
4106.0    3
3217.0    3
5773.0    3
2201.0    3
         ..
146.0     1
5622.0    1
2659.0    1
4409.0    1
2080.0    1
Name: count, Length: 976, dtype: int64


f) Industrial_Use_MCM:

* Info: Water consumption by industries in MCM.

* Use: Helps measure water use efficiency in industries, spot trends by economic activity.

* Missing Values: ~100 rows.

In [33]:
print(dataset["Industrial_Use_MCM"].value_counts())

Industrial_Use_MCM
28.0      3
438.0     3
1562.0    3
555.0     3
1084.0    3
         ..
2646.0    1
2542.0    1
1291.0    1
1936.0    1
168.0     1
Name: count, Length: 915, dtype: int64


g) Domestic_Use_MCM: 

* Info: Water consumed by households (domestic usage) in MCM.

* Use: Important for analyzing urban/rural supply-demand and human needs.

* Missing Values: ~110 rows.

In [34]:
print(dataset["Domestic_Use_MCM"].value_counts())

Domestic_Use_MCM
515.0     5
930.0     4
1943.0    4
1106.0    4
235.0     3
         ..
1387.0    1
995.0     1
606.0     1
625.0     1
1765.0    1
Name: count, Length: 831, dtype: int64


h) Rainfall_mm: 

* Info: Rainfall received (in millimeters) in that year for the region.

* Use: Correlates rainfall with water availability & groundwater recharge.

* Missing Values: ~80 entries.

In [35]:
print(dataset["Rainfall_mm"].value_counts())

Rainfall_mm
1989.0    4
1016.0    4
2453.0    3
2150.0    3
2274.0    3
         ..
372.0     1
2691.0    1
2278.0    1
2596.0    1
110.0     1
Name: count, Length: 878, dtype: int64


i) Groundwater_Level_m:

* Info: Depth of groundwater in meters (lower means depletion).

* Use: Indicates sustainability and stress on reserves.

* Missing Values: ~150 rows.

In [36]:
print(dataset["Groundwater_Level_m"].value_counts())

Groundwater_Level_m
2.52     3
16.71    3
44.86    3
41.21    2
13.60    2
        ..
27.49    1
42.63    1
24.53    1
19.29    1
47.58    1
Name: count, Length: 974, dtype: int64


j) Scarcity_Index: 

* Info: A computed index (0–1 approx) showing scarcity severity. Higher ≈ more scarcity.

* Use: Composite measure to quickly identify stressed regions over time.

* Missing Values: ~90 rows.

In [37]:
print(dataset["Scarcity_Index"].value_counts())

Scarcity_Index
0.59    19
0.41    18
0.13    18
0.86    17
0.66    16
        ..
0.89     6
0.00     5
1.00     4
0.63     3
0.84     2
Name: count, Length: 101, dtype: int64


k) Per_Capita_Consumption_LPD:

* Info: Daily Per Capita water consumption measured in Litres Per Day (LPD).

* Use: Allows direct comparison of human usage habits across regions/time.

* Missing Values: ~100 rows.


In [38]:
print(dataset["Per_Capita_Consumption_LPD"].value_counts())

Per_Capita_Consumption_LPD
225.0    9
405.0    7
353.0    7
156.0    7
466.0    7
        ..
498.0    1
302.0    1
203.0    1
423.0    1
112.0    1
Name: count, Length: 400, dtype: int64


### 7. CHECKING FOR DUPLICATES: 

In [39]:
dataset.duplicated().sum()

0

OBSERVATION: 

As we can see, no duplicates were found in our dataset.

### 8. DETECTING OUTLIERS:
The IQR (Interquartile Range) method was used to detect outliers in all numerical columns as part of the non-visual analysis.

Steps:

* Calculate Q1 (25th percentile) and Q3 (75th percentile)

* Compute IQR = Q3 - Q1

* Define lower and upper bounds:

* Lower bound = Q1 − 1.5 × IQR

* Upper bound = Q3 + 1.5 × IQR

Any value outside this range is considered an outlier

In [40]:
num_cols = dataset.select_dtypes(include=['float64', 'int64']).columns
outliers = {}

In [42]:
outliers = {}
for col in num_cols:
    if dataset[col].isnull().all():
        continue  # Skip the columns with all missing values
    Q1 = dataset[col].quantile(0.25)
    Q3 = dataset[col].quantile(0.75)
    IQR = Q3 - Q1
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR
    outlier_values = dataset[(dataset[col] < lower) | (dataset[col] > upper)][col]
    if not outlier_values.empty:
        outliers[col] = outlier_values.tolist()

In [43]:
if outliers:
    for col, vals in outliers.items():
        print(f"\nOutliers in column '{col}':\n{vals}")
else:
      print("No outliers found in the numerical columns.")

No outliers found in the numerical columns.


OBSERVATIONS:

As we can see, there are no outliers present in the dataset, but:

Scarcity_Index > 1 (should theoretically be ≤ 1).

Per_Capita_Consumption_LPD values as low as ~50 and as high as ~495, need checks against realistic bounds.

Some very low water consumption values (<100 MCM) look implausible.

### Summary of the dataset for Phase 1 (Expected findings):

* Columns present: 15 total (Region, Year, Population_Million, …, Climate_Type, Data_Source).

* Data types: Mix of numeric (float64, int64) & categorical (object).

* Missing values: Several columns contain many NULLs (notably Year, Water_Consumption_MCM, Agriculture_Use_MCM, Domestic_Use_MCM).

* Unusual values: Some extreme numbers in water variables + empty Region/Year rows.

* Categories: Regions = {North, South, East, West, Central}; Continents diverse; Climate_Type spans across 4-5 classes; Data sources mainly {Research, Satellite, Survey, Government Report}.

# PHASE-2